backend: Improve/Cleanup Save()

fd0 opened the pull request:
As mentioned in issue [#1560](https://github.com/restic/restic/pull/1560#issuecomment-364689346) this changes the signature for `backend.Save()`. It now takes a parameter of interface type `RewindReader`, so that the backend implementations or our `RetryBackend` middleware can reset the reader to the beginning and then retry an upload operation. 

The `RewindReader` interface also provides a `Length()` method, which is used in the backend to get the size of the data to be saved. This removes several ugly hacks we had to do to pull the size back out of the `io.Reader` passed to `Save()` before. In the `s3` and `rest` backend this is actively used.

New signature:
```go
// Save stores the data from rd under the given handle.
Save(ctx context.Context, h Handle, rd RewindReader) error
```

Ugly hacks removed:
 * https://github.com/restic/restic/blob/2866f3f31ce98b235010dc6c095e67fc915a83f7/internal/backend/backend_retry.go#L51-L63
 * https://github.com/restic/restic/blob/2866f3f31ce98b235010dc6c095e67fc915a83f7/internal/backend/s3/s3.go#L244-L253

ifedorenko commented on internal/backend/backend_retry.go:
```diff
@@ -47,28 +46,9 @@ func (be *RetryBackend) retry(ctx context.Context, msg string, f func() error) e
-		}
-
-		err = be.Backend.Save(ctx, h, rd)
+		err := be.Backend.Save(ctx, h, length, fn)
```
shouldn't the old reader be closed here on retry?

fd0 replied:
Which old reader?

ifedorenko replied:
The API does not specify if `fn` should return a new reader or reset the same reader on retry. We either need to specify, and ideally enforce, that `fn` must always return the same reader, or assume it'll return new reader, in which case something will need to close the old one during retry.

ifedorenko commented on internal/restic/backend.go:
```diff
@@ -20,8 +20,11 @@ type Backend interface {
+	// backend under the given handle. The function fn may be called several
+	// times to retry the operation in case an error occurs. When fn returns an
+	// error, the process is aborted and the error is returned to the caller.
+	Save(ctx context.Context, h Handle, length int, fn func() (io.Reader, error)) error
```
The proposed API does not look right, to be honest. If `fn` provides the content to save, it should provide both the content reader and the content length (and the error).

I also think adding `length` parameter is orthogonal to `fn` callback.

fd0 replied:
Hm. But for the backend, we **always** know beforehand how much data we're going to save, and that's not going to change on the third try. `fn` is just a way to "reset" the reader, so that is left out of the backend code.

ifedorenko replied:
I am nitpicking, but `length` parameter and the reader returned by `fn` represent related features of the content being saved. I think the API will be easier to understand and use correctly if both of these features were accessible the same way.

Personally, I'd introduce `ContentReader` interface with `Length()`, `Read(...)` and `Reset()` methods and define save as:

```go
Save(ctx context.Context, h Handle, content ContentReader) error
```

I think this will remove all the ugliness you want to remove, without the extra noise required to use`fn` parameter.

fd0 replied:
Uh, that is a nice idea! Thank you very much! I'll try it later

ifedorenko reviewed (commented):
I am not sure... 

I think adding `length` parameter is useful and I am surprised more backends don't need it (http PUT/POST is usually much more efficient when content-length is known).

As for `fn` callback, I am not convinced. Personally I think it trades `seeker` ugliness for `Closer` ugliness during retry, but also adds some noise in each backend implementation. But this is mostly matter of personal preferences, at least if you agree `fn` should return both the reader and the length.

fd0 reviewed (commented):
The main idea I had was to give the backend code a way to "reset" the reader. The length is a parameter of `Save` and not returned by `fn` because it will always be constant: For all data we save in the repo we always know the exact size, which is not changing, even on the third try.

I also thought about this interface, it feels a bit awkward though:
```go
Save(ctx context.Context, h Handle, length int, rd io.Reader, rewind func() error) error
```

I don't see any of the referenced "Closer ugliness", can you show it to me?

fd0 commented:
Ah, the other interface (with the `rewind` method) will not work with the limiters, which needs to be wrapped around the reader again on the next try. Hm.

ifedorenko commented:
Why can't limiter implement `rewind` (aka `reset`) method to simply pass through the call to the wrapped content reader? I haven't thought much about limiters to be honest, but assumed they just throttle traffic per unit of time and don't care whether traffic is a result of retry or not.

fd0 commented:
You may be right, it was late yesterday evening. I'll have a look :)

I like the plain and obvious (at least in retrospect) idea to just wrap it all in an interface, e.g.:

```go
type RewindReader interface {
    io.Reader

    // Rewind sets the Reader back to the start of the
    // data, so all data can be read again.
    Rewind() error
    
    // Length returns the complete number of bytes read
    // from the Reader after Rewind() was called.
    Length() int
}
```

I'll try that later. Thanks for the feedback already!

fd0 committed:
backend: Improve Save()

As mentioned in issue [#1560](https://github.com/restic/restic/pull/1560#issuecomment-364689346)
this changes the signature for `backend.Save()`. It now takes a
parameter of interface type `RewindReader`, so that the backend
implementations or our `RetryBackend` middleware can reset the reader to
the beginning and then retry an upload operation.

The `RewindReader` interface also provides a `Length()` method, which is
used in the backend to get the size of the data to be saved. This
removes several ugly hacks we had to do to pull the size back out of the
`io.Reader` passed to `Save()` before. In the `s3` and `rest` backend
this is actively used.

fd0 commented:
I'm done, please take another look! Apart from the new implementations of `RewindReader` for a byte slice and a file (`restic.ByteReader{}` and `restic.FileReader{}`), the code became even simpler. I like it :)

ifedorenko commented on internal/backend/rest/rest.go:
```diff
@@ -114,12 +115,11 @@ func (b *restBackend) Save(ctx context.Context, h restic.Handle, rd io.Reader) (
-	rd = ioutil.NopCloser(rd)
-
-	req, err := http.NewRequest(http.MethodPost, b.Filename(h), rd)
+	req, err := http.NewRequest(http.MethodPost, b.Filename(h), ioutil.NopCloser(rd))
```
Is `ioutil.NopCloser` wrapper still necessary? `RewindReader` does not implement `Closer` interface as far as I can tell, so client shouldn't be able to close the underlying file already. (same question applies to azure and s3).

fd0 replied:
Good question! I think it isn't strictly necessary, for the reason you mentioned (`RewindReader` does not imply that there's a `Close()` method, but there's no reason there couldn't be a `Close()`). But if we, at some point, add some implementation of `RewindReader` which also has a `Close` method, retries will suddenly stop working mysteriously because the Go stdlib HTTP client will close the body reader for e.g. `POST` requests. That's why the `NopCloser` is there.

So, although it's not strictly necessary, I'd like to keep it in to prevent future bugs.

ifedorenko commented on internal/restic/rewind_reader.go:
```diff
@@ -0,0 +1,90 @@
+
+	// Length returns the number of bytes that can be read from the Reader
+	// after calling Rewind.
+	Length() int
```
shouldn't this be int64? I don't think it's reasonable to assume all uploaded files will be smaller than 2G.

fd0 replied:
Good point, I'll change it to `int64`.

ifedorenko commented on internal/restic/rewind_reader.go:
```diff
@@ -0,0 +1,90 @@
+
+// NewFileReader wraps f in a *FileReader.
+func NewFileReader(f io.ReadSeeker) (*FileReader, error) {
+	pos, err := f.Seek(0, io.SeekEnd)
```
Out of curiosity, why use `file.Seek` and not `file.Stat` to find file length?

fd0 replied:
The interface is smaller. We need `Seek` anyway, and it provides a way to find the length of the file, so we just use that.

fd0 replied:
It'd btw also work to pass in a `bytes.Reader`, which has `Seek` but not `Stat`.

ifedorenko reviewed (approved):
Looks good. I did leave couple of minor comments/questions, but ignore them if they don't make sense.

fd0 committed:
Use int64 for the length in the RewindReader

ifedorenko reviewed (approved):
Thank you for the explanation.

fd0 commented:
Merged, thanks for the feedback!
